### PR TITLE
Add useSequencedContentCloseHandler hook to replace OnClickOutsideWrapper HOC

### PR DESCRIPTION
**Pull request by daltojohnso**
Here I'm introducing a new hook that'll eventually totally replace the HOC `OnClickOutsideWrapper`, and I'm adding it to `TippyPopover` to handle outside clicks and Escape key presses.

The `OnClickOutsideWrapper` HOC is used to handle the closing of popovers/modals when a user clicks outside of the popover/modal content or when the user presses the "Escape" key. Previous to this PR, `TippyPopover` popovers couldn't be used in modals because they weren't added to the global array of rendered popovers found in the HOC file. Originally, I intended to try to get away from using anything but the behavior built into tippy but we need to handle  popovers _in_ modals (which don't use tippy) and the sequenced, one-by-one closing of popovers, so it seemed necessary to either rely on `OnClickOutsideWrapper` or rewrite it to be a hook.

I chose to make a new hook. Using a hook instead of an HOC lets us keep the configurable `lazy` prop on `TippyPopover`, and it seems like a decent opportunity to start moving away from some of our code that relies on `ReactDOM.findDOMNode`.

 The functionality here is _mostly_ the same as it was, except that for now I'm seeing if we can get away without implementing:
1. the `setTimeout(..., 0)` calls that wrap some logic in the `OnClickOutsideWrapper` lifecycle methods
2. the `el.parentNode.style.zIndex = popoverStack.length + 2;` hack also found in `OnClickOutsideWrapper`
3. two unused props, `dismissOnEscape` and `dismissOnClickOutside`

I've made it so that `OnClickOutsideWrapper` uses the `RENDERED_POPOVERS` array found in the file for `useSequencedContentCloseHandler`, and I exported a few of the functions used in the hook so that I can use them in `OnClickOutsideWrapper` as well. Besides that, the component should work the same as it did.

Here's a little demo. You can't see the escape presses, but they're what close the two popovers (notably, one that is a tippy popover and one that is a tether popover) at the start of the video:

https://user-images.githubusercontent.com/13057258/153504348-09c34be0-39ce-4d4c-bafc-14f7d0369a1c.mov

### CODE REVIEW

File: frontend/src/metabase/components/OnClickOutsideWrapper.jsx [hunk 1]
```diff
@@ -1,74 +1,55 @@
...
-      }
+      document.addEventListener("keydown", this._handleEvent, false);
+      window.addEventListener("mousedown", this._handleEvent, true);
     }, 0);
   }
 
   componentWillUnmount() {
-    document.removeEventListener("keydown", this._handleKeyPress, false);
-    window.removeEventListener("mousedown", this._handleClick, true);
+    document.removeEventListener("keydown", this._handleEvent, false);
```
Comment (gusaiani) [inline frontend/src/metabase/components/OnClickOutsideWrapper.jsx:39]: We don't seem to use function names prepended by `_` anymore, do we?

If that is correct, maybe in the future we update the names in this file.

**Review by gusaiani — COMMENTED** [review]
👍

**Review by gusaiani — APPROVED** [review]
👍

File: frontend/src/metabase/components/OnClickOutsideWrapper.jsx (hunk 1 above)
Comment (daltojohnso) [reply frontend/src/metabase/components/OnClickOutsideWrapper.jsx:39]: True -- it'll hopefully be deleted entirely soon enough.

**Pull request merged**